src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -304,15 +317,56 @@ struct
           (Error.of_exn (Parent_not_found (`Parent parent_hash, `Target hash)))
     in
     (* 1.a ; b *)
-    let ( `Hash_after_applying _hash
-        , `Ledger_proof _proof
-        , `Updated_staged_ledger staged_ledger ) =
-      Inputs.Staged_ledger.apply ~logger:t.logger
-        (Breadcrumb.staged_ledger parent_node.breadcrumb)
-        (Inputs.External_transition.staged_ledger_diff transition)
-      |> Or_error.ok_exn
+    let staged_ledger = Breadcrumb.staged_ledger parent_node.breadcrumb in
+    let ledger_hash_from_proof p =
+      Inputs.Ledger_proof.statement_target (Inputs.Ledger_proof.statement p)
+    in
+    let blockchain_state_ledger_hash, blockchain_staged_ledger_hash =
+      let blockchain_state =
+        Protocol_state.blockchain_state transition_protocol_state
+      in
+      ( Protocol_state.Blockchain_state.ledger_hash blockchain_state
+      , Protocol_state.Blockchain_state.staged_ledger_hash blockchain_state )
+    in
+    let transitioned_staged_ledger =
+      match
+          Inputs.Staged_ledger.apply ~logger:t.logger staged_ledger
+            (Inputs.External_transition.staged_ledger_diff transition)
+          |> Or_error.ok_exn
+      
+      with
+      | ( `Hash_after_applying staged_ledger_hash
+        , `Ledger_proof proof_opt
+        , `Staged_ledger transitioned_staged_ledger )
+      ->
+        let target_ledger_hash =
+          match proof_opt with
+          | None ->
+              Option.value_map
+                (Inputs.Staged_ledger.current_ledger_proof
+                   transitioned_staged_ledger)
+                ~f:ledger_hash_from_proof
+                ~default:
+                  (Frozen_ledger_hash.of_ledger_hash
+                     (Ledger.Db.merkle_root t.root_snarked_ledger))
+          | Some proof -> ledger_hash_from_proof proof
+        in
+        if
+          Frozen_ledger_hash.equal target_ledger_hash
+            blockchain_state_ledger_hash
+          && Staged_ledger_hash.equal staged_ledger_hash
+               blockchain_staged_ledger_hash
+        then transitioned_staged_ledger
+        else
+          failwith
+            "Snarked ledger hash and Staged ledger hash after applying the \
+             diff does not match blockchain state's ledger hash and staged \
+             ledger hash resp.\n"
```

Comment:
Actually, one other thing I just realized. This should be a punishment, not a failwith.

Reply:
I will create an issue for this and make another pr.